pkg/forward/forward.go

```diff
@@ -281,80 +264,116 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
+			// fetch the source watermark again, we might not get the latest watermark because of publishing delay,
+			// but ideally we should use the latest to determine the IsLate attribute.
+			processorWM = isdf.fetchWatermark.GetWatermark(readMessages[0].ReadOffset)
+			// assign isLate
+			for _, m := range writeMessages {
+				if processorWM.After(m.EventTime) { // Set late data at source level
+					m.IsLate = true
+				}
 			}
 		}
-	}
 
-	// let's figure out which vertex to send the results to.
-	// update the toBuffer(s) with writeMessages.
-	for _, m := range udfResults {
-		// look for errors in udf processing, if we see even 1 error NoAck all messages
-		// then return. Handling partial retrying is not worth ATM.
-		if m.udfError != nil {
-			udfError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Inc()
-			isdf.opts.logger.Errorw("failed to applyUDF", zap.Error(err))
-			// As there's no partial failure, non-ack all the readOffsets
-			isdf.fromBuffer.NoAck(ctx, readOffsets)
-			return
-		}
-		// update toBuffers
-		for _, message := range m.writeMessages {
-			if err := isdf.whereToStep(message, messageToStep, m.readMessage); err != nil {
-				isdf.opts.logger.Errorw("failed in whereToStep", zap.Error(err))
+		// let's figure out which vertex to send the results to.
+		// update the toBuffer(s) with writeMessages.
+		for _, m := range udfResults {
+			// look for errors in udf processing, if we see even 1 error NoAck all messages
+			// then return. Handling partial retrying is not worth ATM.
+			if m.udfError != nil {
+				udfError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Inc()
+				isdf.opts.logger.Errorw("failed to applyUDF", zap.Error(m.udfError))
+				// As there's no partial failure, non-ack all the readOffsets
 				isdf.fromBuffer.NoAck(ctx, readOffsets)
 				return
 			}
+			// update toBuffers
+			for _, message := range m.writeMessages {
+				if err := isdf.whereToStep(message, messageToStep, m.readMessage); err != nil {
+					isdf.opts.logger.Errorw("failed in whereToStep", zap.Error(err))
+					isdf.fromBuffer.NoAck(ctx, readOffsets)
+					return
+				}
+			}
 		}
-	}
 
-	// forward the message to the edge buffer (could be multiple edges)
-	writeOffsets, err := isdf.writeToBuffers(ctx, messageToStep)
-	if err != nil {
-		isdf.opts.logger.Errorw("failed to write to toBuffers", zap.Error(err))
-		isdf.fromBuffer.NoAck(ctx, readOffsets)
-		return
+		// forward the message to the edge buffer (could be multiple edges)
+		writeOffsets, err = isdf.writeToBuffers(ctx, messageToStep)
+		if err != nil {
+			isdf.opts.logger.Errorw("failed to write to toBuffers", zap.Error(err))
+			isdf.fromBuffer.NoAck(ctx, readOffsets)
+			return
+		}
+	} else {
+		writeOffsets, err = isdf.streamMessage(ctx, dataMessages, processorWM)
+		if err != nil {
```

Comment:
log it?

Reply:
Already log all error cases respectively in `streamMessage`, so skipped here, or you want it to move here and have a general error msg for `streamMessage`?

Reply:
Logging in the invoker might be easier and less code, then in the called function only need to do something like `return nil, fmt.Errorf("failed at xxxxx, error: %w", err)`, so that we can locate the error place.